### PR TITLE
Enhancement: Add hook to axios instance setup

### PR DESCRIPTION
**Pull request by bunchesofdonald**
This adds support to the `Api` class to allow passing in a hook that will receive the `axios` instance so that things like axios interceptors (https://axios-http.com/docs/interceptors) can be attached. This is to support the CSRF implementation here: https://github.com/PrefectHQ/prefect/pull/12354

Related to https://github.com/PrefectHQ/prefect/issues/12347

### CODE REVIEW

File: src/utilities/api.ts [hunk 1]
```diff
@@ -28,32 +29,32 @@ import { WorkspaceWorkQueuesApi } from '@/services/WorkspaceWorkQueuesApi'
 
 // We want the return type to be inferred
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-export function createApi(workspaceConfig: WorkspaceApiConfig) {
+export function createApi(workspaceConfig: WorkspaceApiConfig, instanceSetupHook: AxiosInstanceSetupHook | null = null) {
```
Comment (znicholasbrown) [inline src/utilities/api.ts:32]:
```suggestion
export function createApi(workspaceConfig: WorkspaceApiConfig, instanceSetupHook?: AxiosInstanceSetupHook) {
```

File: src/services/Api.ts [hunk 2]
```diff
@@ -31,9 +33,11 @@ export class Api<T extends PrefectConfig = PrefectConfig> {
   protected apiHeaders: MaybeArray<ApiHeaders> = [getPrefectUIHeaders, getAuthorizationHeaders]
   protected apiBaseUrl: ApiBaseUrl = getPrefectBaseUrl
   protected routePrefix: string | undefined
+  protected instanceSetupHook: AxiosInstanceSetupHook | null
```
Comment (znicholasbrown) [inline src/services/Api.ts:36]:
```suggestion
  protected instanceSetupHook: AxiosInstanceSetupHook | undefined
```

File: src/services/Api.ts [hunk 3]
```diff
@@ -31,9 +33,11 @@ export class Api<T extends PrefectConfig = PrefectConfig> {
...
 
-  public constructor(apiConfig: T) {
+  public constructor(apiConfig: T, instanceSetupHook: AxiosInstanceSetupHook | null = null) {
```
Comment (znicholasbrown) [inline src/services/Api.ts:38]:
```suggestion
  public constructor(apiConfig: T, instanceSetupHook?: AxiosInstanceSetupHook) {
```

**Review by znicholasbrown — APPROVED** [review]
1 set of suggestions so callers don't need to pass `null` to the api constructor

**Comment by pleek91** [comment]
My only thought is it might make sense to add this to the `PrefectConfig` as an optional property. That way there's just a single config object to pass. Otherwise LGTM

**Comment by bunchesofdonald** [comment]
> My only thought is it might make sense to add this to the `PrefectConfig` as an optional property. That way there's just a single config object to pass. Otherwise LGTM

I had it there initially, but this felt different than the config itself as it's a hook that's just inteded to be an _extension_ of this class, which felt different than the real 'config' things like baseURL. If you feel strongly I'm happy to change it.

**Pull request merged**